Removed Context element. doesn't affect functionality, also it preven…

ChengyuanZhao opened the pull request:
fixes #115

ChengyuanZhao committed:
Removed Context element. doesn't affect functionality, also it prevented context.xml files from being recognized.

meltsufin commented:
This is effectively a rollback of #76. I don't remember the details, but I suspect there was a good reason for putting it inside a context. We need to understand that part better.

ChengyuanZhao commented:
Ah yea it looks like the documentation says the Manager element (though not the Valve) must be inside a Context and not a Host element.

I will add a context.xml in the same directory as server.xml, which serves as the Default Context, and will be merged with webapp-level context.xmls, from my trying-it out. The problem is that the Datastore session related classes we wrote need to be unpacked into the WEB-INF or something. I'll work it out and update this PR shortly.

ChengyuanZhao commented:
Actually, it looks like what I just proposed is an even further-back thing:

https://github.com/GoogleCloudPlatform/tomcat-runtime/pull/73/files

ChengyuanZhao commented:
Based on the documentation ( https://tomcat.apache.org/tomcat-8.5-doc/class-loader-howto.html) it looks like #73 wasn't necessary. That PR mentions webapp-loaded classes colliding and overriding classes that the server uses, but the documentation says a class loader will only check its parents' classes and optionally only if the setting is set that way. That would mean the server would never be aware of webapps' classes, even if they conflict.

ChengyuanZhao committed:
moving gcp lib to common classloader

cassand commented:
Indeed the Session Manager needs to be defined inside a Context.

The collision occurs over the services that gRPC registers using a [ServiceLoader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html), e.g the gRPC used by the Tomcat components can register the services exposed by the gRPC of the webapp. 

You can easily reproduce this by loading the tomcat-gcp-lib with the common class-loader and creating a webapp which use an old version of gRPC (the trace and distributed-sessions module should be affected).

ChengyuanZhao commented:
@cassand thank you for the details! I'll try it out and see where to go from there.

ChengyuanZhao commented:
@cassand are my below steps enough to recreate? The test-distributed-war webapp is working properly for me:
- I am loading the tomcat gcp lib with common classloader
- added grpc-all version 0.7.0 from maven central as dependency to test-distributed-war webapp. I see that the existing version the server uses is grpc-all 1.4.0, and 0.7.0 is the oldest available. 
- in SessionServlet.java I added:
```
import io.grpc.Status;
```
Just to make sure the 0.7.0 grpc-all was being used. 
- on starting up the image for test-distributed-war I saw the line
`  inflating: /var/lib/tomcat/webapps/ROOT/WEB-INF/lib/grpc-all-0.7.0.jar `

However, making requests to the webapp localhost:8080/session is working fine and I can see the session data stored on Datastore.

cassand commented:
@ChengyuanZhao the service isn't automatically registered when the library is loaded, but during the first usage, you can use a library of [google-cloud-java](https://github.com/GoogleCloudPlatform/google-cloud-java) to quickly test that.

ChengyuanZhao commented:
@cassand Yea I totally see the error now. I am using Datastore 0.3.0 in the webapp and the server is using 1.2.0, and there are NoSuchMethodErrors coming out indeed. Thanks for getting back on this thread so far after the fact, btw :D. I'll try stuff out and see how we can handle both this problem and also the problem from the linked issue. Maybe there will just need to be a compromise....

ChengyuanZhao commented:
If `<Loader delegate="true"/>` element is added inside the webapp's Context, then the server code is fine, and the webapp's code then uses the server's classes.

ChengyuanZhao commented:
Moving tomcat gcp lib to the server loader didn't fix the problem for me actually :S

cassand commented:
I would suspect this to be caused by the Context definition, I observed different behaviors when the context was defined in server.xml or in context.xml.

I think that a strong separation between the webapp and the custom components would be preferable, as we have spotted a bug for gRPC services but other libraries may cause similar bugs.

ChengyuanZhao commented:
@cassand Did you see the problem go away when you moved the Manager element into server.xml's Context element? I'm still seeing the issue despite doing that, and it seems that only one version of the gRPC classes are used regardless of if they're loaded in separate classloaders as parent and child or as a non-parent/child classloaders. I think because of this we're not losing anything by reverting to moving it out of server.xml, and we're also gaining the ability to use webapps' context.xmls, since using a Context in server.xml causes webapps' context.xmls to be ignored.

Then the issue of it using only a single version of gRPC classes can be resolved in a separate issue/PR.

ChengyuanZhao committed:
Merge branch 'master' into gh-115-context-xml

ChengyuanZhao committed:
Merge branch 'master' into gh-115-context-xml

meltsufin commented on tomcat/pom.xml:
```diff
@@ -251,7 +251,7 @@
             </goals>
             <configuration>
-              <outputDirectory>${project.build.directory}/docker-src/tomcat-base/lib-gcp/</outputDirectory>
+              <outputDirectory>${project.build.directory}/docker-src/tomcat-base/lib/</outputDirectory>
```
I would prefer to keep the GCP jars in a separate directory so that they can be easily removed if not needed by child containers.

ChengyuanZhao replied:
Done.

meltsufin commented on tomcat/src/main/tomcat-base/conf/catalina.properties:
```diff
@@ -68,7 +68,7 @@ common.loader="${catalina.base}/lib","${catalina.base}/lib/*.jar","${catalina.ho
 #       ${catalina.base} path or the ${catalina.home} path contains a comma.
 #       Because double quotes are used for quoting, the double quote character
 #       may not appear in a path.
-server.loader="${catalina.base}/lib-gcp/*.jar"
```
add `"${catalina.base}/lib-gcp/*.jar"` to `common.loader`.

ChengyuanZhao replied:
Done.

meltsufin reviewed (changes requested):
As I understand, there is no way to hide server classes from the webapps in Tomcat. However, webapps can load their own versions of the libs if necessary. From reading the discussion, it seems like there is an issue with some GCP libraries where they cannot coexist even in two different classloaders, but we can deal with that separately. Overall, this seems fine as is, except I would like to keep the GCP jars in a separate directory.

ChengyuanZhao committed:
updated README
